File: software/examples/uc/example_all_data.c
// This example is not self-contained.
// It requres usage of the example driver specific to your platform.
// See the HAL documentation.

#include "bindings/hal_common.h"
#include "bindings/bricklet_imu_v3.h"

#define UID "XYZ" // Change XYZ to the UID of your IMU Bricklet 3.0

void check(int rc, const char* msg);

void example_setup(TF_HalContext *hal);
void example_loop(TF_HalContext *hal);


// Callback function for all data callback
static void all_data_handler(TF_IMUV3 *device, int16_t acceleration[3],
                             int16_t magnetic_field[3], int16_t angular_velocity[3],
                             int16_t euler_angle[3], int16_t quaternion[4],
                             int16_t linear_acceleration[3], int16_t gravity_vector[3],
                             int8_t temperature, uint8_t calibration_status,
                             void *user_data) {
	(void)device; (void)user_data; // avoid unused parameter warning

	tf_hal_printf("Acceleration [X]: %d 1/%d m/s²\n", acceleration[0], 100);
	tf_hal_printf("Acceleration [Y]: %d 1/%d m/s²\n", acceleration[1], 100);
	tf_hal_printf("Acceleration [Z]: %d 1/%d m/s²\n", acceleration[2], 100);
	tf_hal_printf("Magnetic Field [X]: %d 1/%d µT\n", magnetic_field[0], 16);
	tf_hal_printf("Magnetic Field [Y]: %d 1/%d µT\n", magnetic_field[1], 16);
	tf_hal_printf("Magnetic Field [Z]: %d 1/%d µT\n", magnetic_field[2], 16);
	tf_hal_printf("Angular Velocity [X]: %d 1/%d °/s\n", angular_velocity[0], 16);
	tf_hal_printf("Angular Velocity [Y]: %d 1/%d °/s\n", angular_velocity[1], 16);
	tf_hal_printf("Angular Velocity [Z]: %d 1/%d °/s\n", angular_velocity[2], 16);
	tf_hal_printf("Euler Angle [Heading]: %d 1/%d °\n", euler_angle[0], 16);
	tf_hal_printf("Euler Angle [Roll]: %d 1/%d °\n", euler_angle[1], 16);
	tf_hal_printf("Euler Angle [Pitch]: %d 1/%d °\n", euler_angle[2], 16);
	tf_hal_printf("Quaternion [W]: %d 1/%d\n", quaternion[0], 16383);
	tf_hal_printf("Quaternion [X]: %d 1/%d\n", quaternion[1], 16383);
	tf_hal_printf("Quaternion [Y]: %d 1/%d\n", quaternion[2], 16383);
	tf_hal_printf("Quaternion [Z]: %d 1/%d\n", quaternion[3], 16383);
	tf_hal_printf("Linear Acceleration [X]: %d 1/%d m/s²\n", linear_acceleration[0], 100);
	tf_hal_printf("Linear Acceleration [Y]: %d 1/%d m/s²\n", linear_acceleration[1], 100);
	tf_hal_printf("Linear Acceleration [Z]: %d 1/%d m/s²\n", linear_acceleration[2], 100);
	tf_hal_printf("Gravity Vector [X]: %d 1/%d m/s²\n", gravity_vector[0], 100);
	tf_hal_printf("Gravity Vector [Y]: %d 1/%d m/s²\n", gravity_vector[1], 100);
	tf_hal_printf("Gravity Vector [Z]: %d 1/%d m/s²\n", gravity_vector[2], 100);
	tf_hal_printf("Temperature: %I8d °C\n", temperature);
	tf_hal_printf("Calibration Status: %I8u\n", calibration_status);
	tf_hal_printf("\n");
}

static TF_IMUV3 imu;

void example_setup(TF_HalContext *hal) {
	// Create device object
	check(tf_imu_v3_create(&imu, UID, hal), "create device object");

	// Register all data callback to function all_data_handler
	tf_imu_v3_register_all_data_callback(&imu,
	                                     all_data_handler,
	                                     NULL);

	// Set period for all data callback to 0.1s (100ms)
	tf_imu_v3_set_all_data_callback_configuration(&imu, 100, false);
}

void example_loop(TF_HalContext *hal) {
	// Poll for callbacks
	tf_hal_callback_tick(hal, 0);
}

File: software/examples/php/ExampleAllData.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletIMUV3.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletIMUV3;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your IMU Bricklet 3.0

// Callback function for all data callback
function cb_allData($acceleration, $magnetic_field, $angular_velocity, $euler_angle,
                    $quaternion, $linear_acceleration, $gravity_vector, $temperature,
                    $calibration_status)
{
    echo "Acceleration [X]: " . $acceleration[0]/100.0 . " m/s²\n";
    echo "Acceleration [Y]: " . $acceleration[1]/100.0 . " m/s²\n";
    echo "Acceleration [Z]: " . $acceleration[2]/100.0 . " m/s²\n";
    echo "Magnetic Field [X]: " . $magnetic_field[0]/16.0 . " µT\n";
    echo "Magnetic Field [Y]: " . $magnetic_field[1]/16.0 . " µT\n";
    echo "Magnetic Field [Z]: " . $magnetic_field[2]/16.0 . " µT\n";
    echo "Angular Velocity [X]: " . $angular_velocity[0]/16.0 . " °/s\n";
    echo "Angular Velocity [Y]: " . $angular_velocity[1]/16.0 . " °/s\n";
    echo "Angular Velocity [Z]: " . $angular_velocity[2]/16.0 . " °/s\n";
    echo "Euler Angle [Heading]: " . $euler_angle[0]/16.0 . " °\n";
    echo "Euler Angle [Roll]: " . $euler_angle[1]/16.0 . " °\n";
    echo "Euler Angle [Pitch]: " . $euler_angle[2]/16.0 . " °\n";
    echo "Quaternion [W]: " . $quaternion[0]/16383.0 . "\n";
    echo "Quaternion [X]: " . $quaternion[1]/16383.0 . "\n";
    echo "Quaternion [Y]: " . $quaternion[2]/16383.0 . "\n";
    echo "Quaternion [Z]: " . $quaternion[3]/16383.0 . "\n";
    echo "Linear Acceleration [X]: " . $linear_acceleration[0]/100.0 . " m/s²\n";
    echo "Linear Acceleration [Y]: " . $linear_acceleration[1]/100.0 . " m/s²\n";
    echo "Linear Acceleration [Z]: " . $linear_acceleration[2]/100.0 . " m/s²\n";
    echo "Gravity Vector [X]: " . $gravity_vector[0]/100.0 . " m/s²\n";
    echo "Gravity Vector [Y]: " . $gravity_vector[1]/100.0 . " m/s²\n";
    echo "Gravity Vector [Z]: " . $gravity_vector[2]/100.0 . " m/s²\n";
    echo "Temperature: $temperature °C\n";
    echo "Calibration Status: " . sprintf("%08b", $calibration_status) . "\n";
    echo "\n";
}

$ipcon = new IPConnection(); // Create IP connection
$imu = new BrickletIMUV3(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// Register all data callback to function cb_allData
$imu->registerCallback(BrickletIMUV3::CALLBACK_ALL_DATA, 'cb_allData');

// Set period for all data callback to 0.1s (100ms)
$imu->setAllDataCallbackConfiguration(100, FALSE);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
